core/trino-main/src/main/java/io/trino/operator/PagesHash.java

```diff
@@ -168,6 +186,63 @@ public int getAddressIndex(int rightPosition, Page hashChannelsPage, long rawHas
         return -1;
     }
 
+    public long[] getAddressIndex(int[] rightPositions, Page hashChannelsPage, long[] rawHashes)
```

Comment:
I feel like some of the changes here require testing (and `PagesHash` does not have tests)